### PR TITLE
[WIP] Document how to add an Operation

### PR DESCRIPTION
**Pull request by oxinabox**
This is WIP, as I am not 100% sure on how to link into the docs yet.

This probably should be combined with the Shape Inference docs, and called "Contributing".
Or put together in a section.
I'll be back for this later.

### CODE REVIEW

**Comment by JobJob** [comment]
Thanks! Been hoping for improved docs for potential contributors.

File: docs/src/addingops.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+
+Once you have identified the operation you want to add, the following steps should be taken.
+(If you like TDD, feel free to bring the tests first)
+
+ 1. Implement the opteration
+     - This normally means adding it to one of the files `/src/ops/`
+     - You can find the operation's "true name", and its input and attrinbute definitions by looking in the [TensorFlow ops list](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/ops.pbtxt)
+     - You can base the implementation on something already there
+     - Don't forget to shift indexs to be 1 based, as tensorflow defaults to zero based.\
```
Comment (JobJob) [inline docs/src/addingops.md]: typo: indexs -> indexes or indices

File: docs/src/addingops.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+     - Export your new op by adding it to the export's list in `/src/TensorFlow.jl`
```
Comment (JobJob) [inline docs/src/addingops.md]: export's ->exports

File: docs/src/addingops.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ 2. Add the doc string.     
+     - You can normally base this directly off the doc for python
+     - You can find the python docs either by finding the `OpName.md` file, or using the summary+description from the afformentioned op list
+     - Don't forget to change the examples to julia, and to reindex them.
+ 3. Implement some tests
+     - You can often base these of the examples from the docs
```
Comment (JobJob) [inline docs/src/addingops.md]: of -> off

File: docs/src/addingops.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+     - the test file name normally matchs to the name of the file from the `/src/ops` directory, but in `/test`
```
Comment (JobJob) [inline docs/src/addingops.md]: matchs -> matches

File: docs/src/addingops.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+     - For bonus points include a test which requires the gradient to be successfully calculated through your new op -- sometime the op works but for one reason or another it's graident does not
```
Comment (JobJob) [inline docs/src/addingops.md]: graident -> gradient

File: docs/src/addingops.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+ 4. Implement shape inference
+     - The op needs to be added to `src/shape_inference.jl`
+     - This is done by `register_shape("true_name")`, where `true_name` is the name from the [TensorFlow ops list](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/ops.pbtxt)
+     - Again, you can often base it off something already their
```
Comment (JobJob) [inline docs/src/addingops.md]: their->there

File: docs/src/addingops.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+     - The docs for the op are helpful in determining the rules for the output shape -- it is often given explictly (if sometimes cryptically)
+     - If really confused you can use your op and check the shapes by running the operators 
+        - eg `size(run(sess, myop(constant(...)))`. 
+        - You can a large enough quantity of these results to write a pile of tests so that if they pass, you are certain the inference is right.
```
Comment (JobJob) [inline docs/src/addingops.md]: You can ??? a large

maybe: collect

File: docs/src/addingops.md (hunk 1 above)
Comment (oxinabox) [inline docs/src/addingops.md]: Operation here is also spelt wrong.
I should edit this in something with spellcheck.

File: docs/src/addingops.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
+TensorFlow has a huge number of operations.
+TensorFlow.jl implements many of them; but their are many more still to be added.
```
Comment (malmaud) [inline docs/src/addingops.md]: ; => ,
their => there

File: docs/src/addingops.md [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+Linking up to the TensorFlow API to implement a new operation is a fun and easy way to contribute.
+
+See also the [TensorFlow guide for writing language bindings.](https://www.tensorflow.org/extend/language_bindings).
+
+
+Once you have identified the operation you want to add, the following steps should be taken.
+(If you like TDD, feel free to bring the tests first.)
```
Comment (malmaud) [inline docs/src/addingops.md]: Spell out TDD (there are probably people unfamiliar with the acronym)

File: docs/src/addingops.md [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+ 1. Implement the operation
```
Comment (malmaud) [inline docs/src/addingops.md]: Be consistent about ending the list elements with a period or not

File: docs/src/addingops.md [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+     - This normally means adding it to one of the files `/src/ops/`
+     - You can find the operation's "true name", and its input and attrinbute definitions by looking in the [TensorFlow ops list](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/ops.pbtxt)
+     - You can base the implementation on something already there
+     - Don't forget to shift indexes to be 1 based, as tensorflow defaults to zero based.\
```
Comment (malmaud) [inline docs/src/addingops.md]: tensorflow => TensorFlow

File: docs/src/addingops.md [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+See also the [TensorFlow guide for writing language bindings.](https://www.tensorflow.org/extend/language_bindings).
+
+
+Once you have identified the operation you want to add, the following steps should be taken.
+(If you like TDD, feel free to bring the tests first.)
+
+ 1. Implement the operation
+     - This normally means adding it to one of the files `/src/ops/`
+     - You can find the operation's "true name", and its input and attrinbute definitions by looking in the [TensorFlow ops list](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/ops.pbtxt)
```
Comment (malmaud) [inline docs/src/addingops.md]: attrinbute => attribute

File: docs/src/addingops.md [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+     - You can base the implementation on something already there
+     - Don't forget to shift indexes to be 1 based, as tensorflow defaults to zero based.\
+     - Export your new op by adding it to the exports list in `/src/TensorFlow.jl`
+ 2. Add the doc string.     
+     - You can normally base this directly off the doc for python
+     - You can find the python docs either by finding the `OpName.md` file, or using the summary+description from the afformentioned op list
+     - Don't forget to change the examples to julia, and to reindex them.
```
Comment (malmaud) [inline docs/src/addingops.md]: julia => Julia

**Pull request merged**